File: backend/GraficaModerna.API/Controllers/AuthController.cs
using System.IdentityModel.Tokens.Jwt;
using System.Security.Claims;
using GraficaModerna.Application.DTOs;
using GraficaModerna.Application.Interfaces;
using GraficaModerna.Domain.Constants;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.AspNetCore.RateLimiting;

namespace GraficaModerna.API.Controllers;

[Route("api/[controller]")]
[ApiController]
public class AuthController(
    IAuthService authService,
    ITokenBlacklistService blacklistService,
    ILogger<AuthController> logger,
    IContentService contentService) : ControllerBase
{
    private readonly IAuthService _authService = authService;
    private readonly ITokenBlacklistService _blacklistService = blacklistService;
    private readonly ILogger<AuthController> _logger = logger;
    private readonly IContentService _contentService = contentService;

    [EnableRateLimiting("AuthPolicy")]
    [HttpPost("register")]
    public async Task<ActionResult> Register(RegisterDto dto)
    {
        var result = await _authService.RegisterAsync(dto);
        SetTokenCookies(result.AccessToken, result.RefreshToken);

        return Ok(new
        {
            result.Email,
            result.Role,
            message = "Cadastro realizado com sucesso."
        });
    }

    [EnableRateLimiting("AuthPolicy")]
    [HttpPost("login")]
    public async Task<ActionResult> Login(LoginDto dto)
    {
        var userLoginDto = dto with { IsAdminLogin = false };
        var result = await _authService.LoginAsync(userLoginDto);

        SetTokenCookies(result.AccessToken, result.RefreshToken);

        return Ok(new
        {
            result.Email,
            result.Role,
            message = "Login realizado com sucesso."
        });
    }

    [HttpPost("logout")]
    [Authorize]
    public async Task<IActionResult> Logout()
    {
        string? token = null;
        if (Request.Cookies.TryGetValue("jwt", out var cookieToken))
            token = cookieToken;
        else if (Request.Headers.TryGetValue("Authorization", out var authHeader))
        {
            var header = authHeader.ToString();
            if (header.StartsWith("Bearer ", StringComparison.OrdinalIgnoreCase))
                token = header["Bearer ".Length..].Trim();
        }

        if (!string.IsNullOrEmpty(token))
            try
            {
                var handler = new JwtSecurityTokenHandler();
                if (handler.CanReadToken(token))
                {
                    var jwtToken = handler.ReadJwtToken(token);
                    await _blacklistService.BlacklistTokenAsync(token, jwtToken.ValidTo);
                }
            }
            catch (Exception ex)
            {
                _logger.LogWarning("Erro ao processar blacklist no logout: {Message}", ex.Message);
            }

        Response.Cookies.Delete("jwt");
        Response.Cookies.Delete("refreshToken");

        return Ok(new { message = "Deslogado com sucesso" });
    }

    [HttpGet("check-auth")]
    [AllowAnonymous]
    public IActionResult CheckAuth()
    {
        if (User.Identity?.IsAuthenticated != true)
        {
            return Ok(new { isAuthenticated = false, role = (string?)null });
        }
        var role = User.FindFirstValue(ClaimTypes.Role);
        return Ok(new { isAuthenticated = true, role });
    }

    [HttpGet("profile")]
    [Authorize(Roles = Roles.User + "," + Roles.Admin)]
    public async Task<ActionResult<UserProfileDto>> GetProfile()
    {
        var userId = User.FindFirstValue(ClaimTypes.NameIdentifier);
        if (string.IsNullOrEmpty(userId)) return Unauthorized();

        var profile = await _authService.GetProfileAsync(userId);
        return Ok(profile);
    }

    [HttpPut("profile")]
    [Authorize(Roles = Roles.User + "," + Roles.Admin)]
    public async Task<IActionResult> UpdateProfile(UpdateProfileDto dto)
    {
        var settings = await _contentService.GetSettingsAsync();
        if (settings.TryGetValue("purchase_enabled", out var enabled) && enabled == "false")
        {
            var role = User.FindFirstValue(ClaimTypes.Role);
            if (role != Roles.Admin)
            {
                return StatusCode(403, new { message = "Edição de perfil desativada temporariamente." });
            }
        }

        var userId = User.FindFirstValue(ClaimTypes.NameIdentifier);
        if (string.IsNullOrEmpty(userId)) return Unauthorized();

        await _authService.UpdateProfileAsync(userId, dto);
        return NoContent();
    }

    [HttpPost("refresh-token")]
    [AllowAnonymous]
    public async Task<IActionResult> RefreshToken()
    {
        var accessToken = Request.Cookies["jwt"];
        var refreshToken = Request.Cookies["refreshToken"];

        if (string.IsNullOrEmpty(accessToken) || string.IsNullOrEmpty(refreshToken))
            return BadRequest("Tokens não encontrados nos cookies.");

        var tokenModel = new TokenModel(accessToken, refreshToken);

        var result = await _authService.RefreshTokenAsync(tokenModel);
        SetTokenCookies(result.AccessToken, result.RefreshToken);

        return Ok(new { result.Email, result.Role });
    }

    [HttpPost("confirm-email")]
    [AllowAnonymous]
    public async Task<IActionResult> ConfirmEmail([FromBody] ConfirmEmailDto dto)
    {
        await _authService.ConfirmEmailAsync(dto);
        return Ok(new { Message = "Email confirmado com sucesso!" });
    }

    [HttpPost("forgot-password")]
    [AllowAnonymous]
    public async Task<IActionResult> ForgotPassword([FromBody] ForgotPasswordDto dto)
    {
        await _authService.ForgotPasswordAsync(dto);
        return Ok(new { Message = "Se o e-mail estiver cadastrado, você receberá um link de recuperação." });
    }

    [HttpPost("reset-password")]
    [AllowAnonymous]
    public async Task<IActionResult> ResetPassword([FromBody] ResetPasswordDto dto)
    {
        await _authService.ResetPasswordAsync(dto);
        return Ok(new { Message = "Senha redefinida com sucesso. Faça login com a nova senha." });
    }

    private void SetTokenCookies(string accessToken, string refreshToken)
    {
        var cookieOptions = new CookieOptions
        {
            HttpOnly = true,
            Secure = true,
            SameSite = SameSiteMode.Lax,
            Expires = DateTime.UtcNow.AddMinutes(15)
        };

        var refreshCookieOptions = new CookieOptions
        {
            HttpOnly = true,
            Secure = true,
            SameSite = SameSiteMode.Lax,
            Expires = DateTime.UtcNow.AddDays(7)
        };

        Response.Cookies.Append("jwt", accessToken, cookieOptions);
        Response.Cookies.Append("refreshToken", refreshToken, refreshCookieOptions);
    }
}

File: grafica-moderna/app/Http/Controllers/AuthController.php
<?php

namespace App\Http\Controllers;

use Illuminate\Http\Request;
use Illuminate\Support\Facades\Cookie;
use Illuminate\Support\Facades\Log;

use App\Services\AuthService;
use App\Services\TokenBlacklistService;
use App\Services\ContentService;
use App\Http\Requests\Auth\RegisterRequest;
use App\Http\Requests\Auth\LoginRequest;
use App\Http\Requests\Auth\UpdateProfileRequest;

use Firebase\JWT\JWT;
use Firebase\JWT\Key;

class AuthController extends Controller
{
    protected AuthService $authService;
    protected TokenBlacklistService $blacklistService;
    protected ContentService $contentService;

    public function __construct(
        AuthService $authService,
        TokenBlacklistService $blacklistService,
        ContentService $contentService
    ) {
        // Rate Limiting aplicado via middleware no bootstrap ou rotas
        $this->authService = $authService;
        $this->blacklistService = $blacklistService;
        $this->contentService = $contentService;
    }

    // ======================================================
    // REGISTER
    // ======================================================

    public function register(RegisterRequest $request)
    {
        // Validação automática via RegisterRequest
        $data = $request->validated();

        $result = $this->authService->register($data);

        $this->setTokenCookies(
            $result->accessToken,
            $result->refreshToken
        );

        return response()->json([
            'email'   => $result->email,
            'role'    => $result->role,
            'message' => 'Cadastro realizado com sucesso.'
        ]);
    }

    // ======================================================
    // LOGIN (USER)
    // ======================================================

    public function login(LoginRequest $request)
    {
        // Validação automática via LoginRequest
        $data = $request->validated();
        $data['isAdminLogin'] = false;

        $result = $this->authService->login($data);

        $this->setTokenCookies(
            $result->accessToken,
            $result->refreshToken
        );

        return response()->json([
            'email'   => $result->email,
            'role'    => $result->role,
            'message' => 'Login realizado com sucesso.'
        ]);
    }

    // ======================================================
    // LOGOUT
    // ======================================================

    public function logout(Request $request)
    {
        $token = $request->cookie('jwt') ?? $request->bearerToken();

        if ($token) {
            try {
                $decoded = JWT::decode(
                    $token,
                    new Key(config('app.jwt_secret'), 'HS256')
                );

                if (isset($decoded->exp)) {
                    $this->blacklistService->blacklist(
                        $token,
                        $decoded->exp
                    );
                }
            } catch (\Exception $e) {
                Log::warning(
                    'Erro ao processar blacklist no logout',
                    ['message' => $e->getMessage()]
                );
            }
        }

        return response()->json([
            'message' => 'Deslogado com sucesso'
        ])
        ->withCookie(cookie()->forget('jwt'))
        ->withCookie(cookie()->forget('refreshToken'));
    }

    // ======================================================
    // CHECK AUTH
    // ======================================================

    public function checkAuth(Request $request)
    {
        if (!$request->user()) {
            return response()->json([
                'isAuthenticated' => false,
                'role' => null
            ]);
        }

        return response()->json([
            'isAuthenticated' => true,
            'role' => $request->user()->role
        ]);
    }

    // ======================================================
    // PROFILE
    // ======================================================

    public function getProfile(Request $request)
    {
        if (!$request->user()) {
            return response()->json([], 401);
        }

        return response()->json(
            $this->authService->getProfile($request->user()->id)
        );
    }

    public function updateProfile(UpdateProfileRequest $request)
    {
        $settings = $this->contentService->getSettings();

        if (
            isset($settings['purchase_enabled']) &&
            $settings['purchase_enabled'] === 'false'
        ) {
            if ($request->user()->role !== 'Admin') {
                return response()->json([
                    'message' => 'Edição de perfil desativada temporariamente.'
                ], 403);
            }
        }

        // Validação automática via UpdateProfileRequest
        $data = $request->validated();

        $this->authService->updateProfile(
            $request->user()->id,
            $data
        );

        return response()->noContent();
    }

    // ======================================================
    // REFRESH TOKEN
    // ======================================================

    public function refreshToken(Request $request)
    {
        $accessToken  = $request->cookie('jwt');
        $refreshToken = $request->cookie('refreshToken');

        if (!$accessToken || !$refreshToken) {
            return response()->json([
                'message' => 'Tokens não encontrados nos cookies.'
            ], 400);
        }

        $result = $this->authService->refreshToken([
            'accessToken'  => $accessToken,
            'refreshToken' => $refreshToken
        ]);

        $this->setTokenCookies(
            $result->accessToken,
            $result->refreshToken
        );

        return response()->json([
            'email' => $result->email,
            'role'  => $result->role
        ]);
    }

    // ======================================================
    // EMAIL / PASSWORD
    // ======================================================

    public function confirmEmail(Request $request)
    {
        // Validação simples inline para casos menores
        $data = $request->validate([
            'userId' => 'required|string',
            'token'  => 'required|string'
        ]);

        $this->authService->confirmEmail($data);
        return response()->json([
            'message' => 'Email confirmado com sucesso!'
        ]);
    }

    public function forgotPassword(Request $request)
    {
        $data = $request->validate(['email' => 'required|email']);
        
        $this->authService->forgotPassword($data);
        
        return response()->json([
            'message' => 'Se o e-mail estiver cadastrado, você receberá um link de recuperação.'
        ]);
    }

    public function resetPassword(Request $request)
    {
        $data = $request->validate([
            'userId' => 'required|string',
            'token' => 'required|string',
            'newPassword' => 'required|min:6|confirmed'
        ]);

        $this->authService->resetPassword($data);
        
        return response()->json([
            'message' => 'Senha redefinida com sucesso. Faça login com a nova senha.'
        ]);
    }

    // ======================================================
    // HELPERS
    // ======================================================

    private function setTokenCookies(string $access, string $refresh): void
    {
        Cookie::queue(
            cookie('jwt', $access, 15, null, null, true, true, false, 'Lax')
        );

        Cookie::queue(
            cookie('refreshToken', $refresh, 60 * 24 * 7, null, null, true, true, false, 'Lax')
        );
    }
}